File: 0637-average-of-levels-in-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) 
    {
        if(root == nullptr)
            return vector<double>();
        
        queue<TreeNode*> lev;
        vector<double> result;
        lev.push(root);
        
        while(!lev.empty())
        {
            double count = 0;
            double sum = 0;
            int size = lev.size();
            
            for(int i = 0; i < size; i++)
            {
                TreeNode* cur_node = lev.front();
                sum += cur_node->val;
                count++;
                
                if(cur_node->left != nullptr || cur_node->right != nullptr)
                {
                    if(cur_node->left != nullptr)
                    {
                        lev.push(cur_node->left);
                       
                    }
                    if(cur_node->right != nullptr)
                    {
                        lev.push(cur_node->right);
        
                    }
                }
                lev.pop();
            }
            if(size > 0)
            {
                result.push_back(sum/count);
            }
        }
        return result;
        
    }
};

File: 0637-average-of-levels-in-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        
        Q   = collections.deque()
        ans = []
        
        
        Q.append(root)
        
        while Q:
            avg  = 0.0
            n    = len(Q)
            
            nextLevel = collections.deque()
            
            while Q:
                node = Q.popleft()
                avg  += node.val
                
                if node.left:
                    nextLevel.append(node.left)
                if node.right:
                    nextLevel.append(node.right)
                    
            avg /= n
            ans.append(avg)
            Q = nextLevel
        
        return ans
        
